src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/CacheCustomMetricsCollector.java

```diff
@@ -0,0 +1,126 @@
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics.CacheCustomDimension;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.AllMetrics.CacheCustomValue;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsConfiguration;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.MetricsProcessor;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.metrics.PerformanceAnalyzerMetrics;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.indices.IndicesService;
+
+/*
+ * Unlike Cache Hit, Miss, Eviction Count and Size, which is tracked on a per shard basis,
+ * the Cache Max size is a node-level static setting and thus, we need a custom collector
+ * (other than NodeStatsMetricsCollector which collects the per shard metrics) for this
+ * metric.
+ *
+ * CacheCustomMetricsCollector collects the max size for the Field Data and Shard Request
+ * Cache currently and can be extended for remaining cache types and any other node level
+ * cache metric.
+ *
+ */
+public class CacheCustomMetricsCollector extends PerformanceAnalyzerMetricsCollector implements MetricsProcessor {
+    public static final int SAMPLING_TIME_INTERVAL = MetricsConfiguration.CONFIG_MAP.get(
+            CacheCustomMetricsCollector.class).samplingInterval;
+    private static final int KEYS_PATH_LENGTH = 0;
+    private StringBuilder value;
+
+    public CacheCustomMetricsCollector() {
+        super(SAMPLING_TIME_INTERVAL, "CacheCustomMetrics");
+        value = new StringBuilder();
+    }
+
+    @Override
+    public void collectMetrics(long startTime) {
+        IndicesService indicesService = ESResources.INSTANCE.getIndicesService();
+        if (indicesService == null) {
+            return;
+        }
+
+        value.setLength(0);
+        value.append(PerformanceAnalyzerMetrics.getJsonCurrentMilliSeconds());
+        // This is for backward compatibility. Core ES may or may not emit custom shard metrics
+        // (depending on whether the patch has been applied or not). Thus, we need to use reflection
+        // to check whether getMaxWeight() method exist in Cache.java
+        //
+        // Currently, we are collecting maxWeight metrics only for FieldData and Shard Request Cache.
+        CacheMaxSizeStatus fieldDataCacheMaxSizeStatus = AccessController.doPrivileged(
+                (PrivilegedAction<CacheMaxSizeStatus>) () -> {
+                    try {
+                        Cache fieldDataCache = indicesService.getIndicesFieldDataCache().getCache();
+                        long fieldDataMaxSize = (Long) FieldUtils.readField(fieldDataCache, "maximumWeight", true);
```

Comment:
should we create a dedicated Const.java and define the string constants used by PerformanceController in it ?

Reply:
Moved them to a const file under Decision maker.